build_wheels.py

```diff
@@ -0,0 +1,113 @@
+        if float(branch.split('v')[1]) >= 5.0:
+            idf_branches.append(branch)
+idf_branches.append('master')
+
+# Constraints files versions list
+idf_constraints = []
+for branch in idf_branches:
+    try:
+        splitted_branch = branch.split('release/')
+        idf_constraints.append(splitted_branch[1])
+    except IndexError:
+        if AUTO_VERSION_CONSTRAINS_FILE:
+            idf_last_release = idf_constraints[-1].split('v')[1].split('.')
+            idf_master_minor = str(int(idf_last_release[1])+1)
+            if requests.get(f'{idf_constraints_url}v{idf_last_release[0]}.{idf_master_minor}.txt').status_code == 200:
+                idf_constraints.append(f'v{idf_last_release[0]}.{idf_master_minor}')
+            else:
+                idf_constraints.append(idf_constraints[-1])
+        else:
+            idf_constraints.append('6.0')
+
+print(f'ESP-IDF branches to be downloaded requirements for:\n{idf_branches}\n')
+print(f'ESP-IDF constrains files versions to be downloaded requirements for:\n{idf_constraints}\n')
+
+idf_resources_url = 'https://raw.githubusercontent.com/espressif/esp-idf/'
+requirements_txt = []
+requirements_set = set()
+
+# --- Download all requirements from all the branches requirements and constraints files --- #
+for i, branch in enumerate(idf_branches):
+    idf_requirements_json_url = f'{idf_resources_url}{branch}/tools/requirements.json'
+    constraint_file_url = f'https://dl.espressif.com/dl/esp-idf/espidf.constraints.{idf_constraints[i]}.txt'
+
+    r = requests.get(idf_requirements_json_url)
+    if r.status_code == 200:
+        idf_requirements_json = json.loads(r.content)
+    else:
+        print(f'\nFailed to download requirements JSON for branch {branch}')
+        continue
+
+    # Download requirements files from all requirements groups specified in JSON
+    print(f'---------- ESP-IDF BRANCH {branch} ----------')
+    for feature in idf_requirements_json['features']:
+        r = requests.get(f"{idf_resources_url}{branch}/{feature['requirement_path']}")
+        if r.status_code == 200:
+            requirements_txt += r.text.splitlines()
+            print(f"Added ESP-IDF {feature['name']} requirements")
+        else:
+            print(f"Failed to download feature (requirement group) '{feature['name']}'")
+
+    # Download constraints file for specific branch
+    r = requests.get(constraint_file_url)
+    if r.status_code == 200:
+        requirements_txt += r.text.splitlines()
+        print(f'Added ESP-IDF constraints file {idf_constraints[i]} for branch {branch}')
+    else:
+        print(f'Failed to download ESP-IDF constraints file {idf_constraints[i]} for branch {branch}')
+
+    # Add requirements into set to prevent duplicates
+    for line in map(str.strip, requirements_txt):
+        if line != '\n' or line != '\r\n' or line != '\r':
+            # check if in the line or the line itself is not a comment
+            if '#' in line:
```

Comment:
1. This check will go through the whole line and will check each character. (for all lines)
2. If match has been found then it will go again essentially doing the same and splitting.

So you can do the following for every line
```python
line = line.split('#', 1)[0]
```
which at least saves no 2. and makes a code simpler by removing one "if" branch.

Please let me know if these kind of optimizations doesn't make sense to you. I can explain more.